use jsonlite for data and init files

rok-cesnovar opened the pull request:
This replaces the use of stan_rdump with the use of json with the jsonlite package.

I did a performance test of rjson, RJSONIO and jsonlite and rjson and jsonlite have similar performance, but rjson isnt really configurable and behaves weird with 2D arrays, etc.

This is my first try of doing a PR for any R package so please let me know if I messed up.

There is an open question whether we should only allow the use of json for 2.21+ as that is where it will actually be faster.

jgabry commented:
Awesome, thanks @rok-cesnovar! I will review this as soon as possible.

jgabry commented:
Also, going forward feel free to just create new branches and make PRs that way instead of via a fork. Fine to do it from the fork, just letting you know.

rok-cesnovar commented on DESCRIPTION:
```diff
@@ -21,7 +21,8 @@ SystemRequirements: CmdStan <https://mc-stan.org/users/interfaces/cmdstan>
     processx,
-    R6
+    R6,
+    jsonlite
```
I would say the minimum is 1.2, since that is when `write_json` was introduced (I am going to switch to that as opposed to toJSON with write. Will add that requirement.

rok-cesnovar commented on R/utils.R:
```diff
@@ -172,24 +172,17 @@ repair_variable_names <- function(names) {
-  )
+  temp_file <- tempfile(pattern = "standata-", fileext = ".json")
+  # write the data in the JSON file with full precision
+  json_data <- jsonlite::toJSON(data, auto_unbox = TRUE, digits = NA)
```
Yes, this should be as convenient as rdump. That is a given. I just wasnt aware of the convenience features stan_rdump has.

jgabry replied:
No problem, I forgot about them or I would have mentioned this sooner, sorry!

rok-cesnovar commented:
> Also, going forward feel free to just create new branches and make PRs that way instead of via a fork. Fine to do it from the fork, just letting you know.

WIll do. I just default to using a fork out of fear of breaking anything :) Will keep the fork for any experimenting and such and do these smaller PRs on the main repo. Thanks for the heads up.

I will fix the convenience stuff in this PR. I wasnt aware of those features because I didnt read the docs of stan_rdump carefully enough, sorry about that.

jgabry commented:
> > Also, going forward feel free to just create new branches and make PRs that way instead of via a fork. Fine to do it from the fork, just letting you know.
> WIll do. I just default to using a fork out of fear of breaking anything :) Will keep the fork for any experimenting and such and do these smaller PRs on the main repo. Thanks for the heads up.

I know the feeling, but right now as long as you're not on the master branch you won't be able to break anything!

> I will fix the convenience stuff in this PR. I wasnt aware of those features because I didnt read the docs of stan_rdump carefully enough, sorry about that.

No worries at all. Like I said above I forgot this didn't occur to me until trying out this PR. If I had remembered I would have given you a heads up about that.

rok-cesnovar commented:
I think I was able to replicate what `stan_rdump` does. For now I made a `stan_jsondump` function that resembles the `stan_rdump` in terms of the signature, except that `quiet` and `width` arent supported arguments.

These are all the test arguments I tried out.
```R
test_factor = factor(sample(0:2, 15, replace = TRUE), labels = c("c1", "c2", "c3"))
test_bool <- c(TRUE, FALSE, TRUE, FALSE, TRUE, TRUE, TRUE, TRUE, FALSE,FALSE,FALSE)
# this should all be floats
test_mixed <- c(1.0, 2.0, 3, 4, 5, 6, 3.14, 2.782)
# this should all be integer (should ignore .0)
test_integer <- c(1.0, 2.0, 3, 4, 5, 6, 7, 8, 9, 10)
test_overflow_int <- c(.Machine$integer.max,.Machine$integer.max+1)
test_extreme_double <- c(.Machine$double.xmin, .Machine$double.xmax)
test_extreme_double2 <- c(.Machine$double.xmin, 1.6976932233333e+308)
# this should be converted to data.matrix
test_data_frame <- read.table('https://raw.github.com/wiki/stan-dev/rstan/rats.txt', header = TRUE)
# tests lists
I <- 4; J <- 5; K <- 10
test_list <- lapply(1:I, function(i) rnorm(J))
test_list2 <- a <- lapply(1:I, function(i) array(rnorm(J * K), dim = c(J, K)))
standata <- list(N = 10,
                 test_factor = test_factor,
                 test_bool = test_bool,
                 test_mixed = test_mixed,
                 test_integer = test_integer,
                 test_overflow_int = test_overflow_int,
                 test_extreme_double = test_extreme_double,
                 test_extreme_double2 = test_extreme_double2,
                 test_data_frame = test_data_frame,
                 test_list = test_list,
                 test_list2 = test_list2)

file_rdump = tempfile()
file_jsondump = tempfile()

stan_rdump(list = names(standata), envir = as.environment(standata), file = file_rdump)
stan_jsondump(list = names(standata), envir = as.environment(standata), file = file_jsondump)

rdump <- read_rdump(file_rdump)
jsondump <- read_json(file_jsondump, simplifyVector = TRUE)

print(readLines(file_rdump))
print(readLines(file_jsondump))

print(match_lists(rdump, jsondump))
```
If I output them to a file and read them back in the lists match. The list match function is just: 

```R
match_lists <- function (list1, list2) {
  if(length(names(list1)) != length(names(list2))) {
    return(FALSE)
  }
  for(var1_name in names(list1)) {
    if(is.null(list2[[var1_name]])) {
      warning(paste(var1_name, " not found!"))
      return(FALSE)
    }
    if(!all(rdump$N == jsondump$N)) {
      warning(paste("Values in ",var1_name, " dont match!"))
      return(FALSE)
    }
  }
  TRUE
}
```
Going to push this on the branch. The only thing I want to try out is this lists thing with an actual model. Do you know of any example models that use this?

rok-cesnovar commented:
I added a few more tests, now I think this is good for another look. 
The cmdstanr_jsondump resembles the what stan_rdump does except:
  - always writes to file
  - all variables must exist in the environment (warning if they dont)
  - no check if the names are valid Stan names (that would get handled by Cmdstan)
  - no quiet mode

jgabry commented:
Awesome, thanks! I’ll take a look ASAP.

jgabry commented on R/utils.R:
```diff
@@ -167,34 +167,70 @@ repair_variable_names <- function(names) {
+  if (!is.character(file) || !nzchar(file)) {
+    warning("The supplied filename is invalid!")
+    return(invisible(character()))
+  }
```
I know stan_rdump does warnings and returns empty character() objects, but unless you like that, I'd prefer just throwing an error here. That's probably what stan_rdump should be doing too but it's a really old function and hasn't been refactored in ages.

rok-cesnovar replied:
Tried to resemble the behaviour of rstan again here. Will change back. Agree with throwing error here.

jgabry commented on R/utils.R:
```diff
@@ -167,34 +167,70 @@ repair_variable_names <- function(names) {
+  # check if all variables in list exist
+  variable_exists <- sapply(list, exists, envir = envir)
+  if(!all(variable_exists)) {
+    warning(paste("The following objects were not found: ", paste(list[!variable_exists], collapse = ", ")))
```
Same. If you don't have a preference then I'd go with error here.

rok-cesnovar replied:
Agree with throwing error. This one is actually being removed because of the move to the new signature.

jgabry commented on R/utils.R:
```diff
@@ -167,34 +167,70 @@ repair_variable_names <- function(names) {
+  # call to write JSON with
+  # unboxing variables (N = 10 is stored as N : 10, not N: [10])
+  # handling factors as integers
+  jsonlite::write_json(data, auto_unbox = TRUE, factor = "integer", path = file)
```
`jsonlite::write_data()` returns `NULL` if you assign it to a value like 

```
x <- jsonlite::write_json(...) # x is NULL
```

which is fine, but currently the doc for `cmdstanr_jsondump` says it returns a file path (looks just like a copy and paste issue). 

Assuming `cmdstan_jsondump()` is exposed to the user I don't really know if it makes sense to return anything to the user from this function or not. `stan_rdump()` just returns the names of the objects it wrote I think, but I don't know if that's super useful. And we don't need it to return a file path since it's just the value of the `file` argument. So I guess `NULL` is fine, just change the doc? What do you think?

rok-cesnovar replied:
Yes, a copy paste issue. Will fix.

jgabry commented:
> The only thing I want to try out is this lists thing with an actual model. Do you know of any example models that use this?

I can't think of a real one at the moment (although there definitely are), but I think you could test with something like this: 

```stan
data {
  int I; 
  int J;
  int K;
  matrix[J,K] x[I]; 
}
parameters {
  real mu;
}
model {
  mu ~ normal_std();
  for (i in 1:I) to_vector(x[i]) ~ normal(mu, 1);
}
```

```r
x <- list()
for (i in 1:10) {
  x[[i]] <- matrix(rnorm(15), nrow = 5, ncol = 3)
}

standata <- list(x = x) # then pass this to cmdstanr 
```

jgabry commented on R/utils.R:
```diff
@@ -167,34 +167,70 @@ repair_variable_names <- function(names) {
+#' @return Path to temporary file containing the data.
+cmdstanr_jsondump <- function(list,
+                          file,
+                          envir = parent.frame()) {
```
Also, for the exported function what do you think of the name `write_stan_json()` instead of `cmdstanr_jsondump()`? I don't have a strong preference so up to you, just throwing it out there as an option.

rok-cesnovar replied:
Yeah, I just presumed that we would want the same signature as stan_rdump. I agree that its not intuitive. Will change back, no problem at all! Aslo prefer write_stan_json much better. Again wanted to resemble stan_rdump.

jgabry replied:
Cool. I think once cmdstanr is released  it probably actually makes sense for us to deprecate `rstan::stan_rdump` and put a version of it into cmdstanr called `write_stan_rdump` with the same signature as the new `write_stan_json`. Really the only reason rstan needed to have `stan_rdump` was because an R package for all things cmdstan didn’t exist at the time.

jgabry commented:
Also once cmdstanr is released I should add a message in `rstan::stan_rdump` suggesting using cmdstanr’s json writer if json is the best method to pass data to cmdstan going forward.

rok-cesnovar commented on tests/testthat/test-json.R:
```diff
@@ -0,0 +1,81 @@
+test_that("test JSON output integer vector", {
+  skip_on_cran()
+  temp_file <- tempfile()
+  N = c(1.0, 2.0, 3, 4)
```
Ah ok, that one is weird. 

Not sure if this R fact applies to this jsonlite case. `c(1.0, 2.0, 3, 4)` is written to json as a vector of integers. This is the same thing that stan_rdump does.

rok-cesnovar commented on DESCRIPTION:
```diff
@@ -23,7 +23,8 @@ SystemRequirements: CmdStan <https://mc-stan.org/users/interfaces/cmdstan>
     processx,
-    R6
+    R6,
+    jsonlite (>= 1.2)
```
Ah, yes. That makes sense. Will fix.

rok-cesnovar commented:
I think I addressed all the review comments. Thank you for the thorough review. 

I had to refactor the handling of the special list argument that `stan_rdump` handles. I tested it thoroughly now and cmdstanr on master and on this branch draw the exact same samples for the model you posted.
